### PR TITLE
Allow neon to compile from the source

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo-devpack-dotnet/issues/132

- This PR Allow to compile c# and VB scripts from it source

Without arguments

```
Neo.Compiler.MSIL 2.4.1
2015-2019 The Neo Project

ERROR(S):
  Required option 'f, file' is missing.

  -f, --file          Required. Specify the file to compile to NEF file

  -r, --references    Compilation references

  --help              Display this help screen.

  --version           Display version information.
``` 

Compile from source

`-f "V:\Temp\sc.cs" -r "Neo.SmartContract.Framework.dll"`

```
Neo.Compiler.MSIL console app v2.4.1.0
Auto Insert entrypoint.
convert succ
gen abi succ
write:sc.nef
write:sc.abi.json
write:sc.manifest.json
SUCC
```

Compile as usual

`-f "V:\Temp\sc.dll"`

```
Neo.Compiler.MSIL console app v2.4.1.0
Auto Insert entrypoint.
convert succ
gen abi succ
write:sc.nef
write:sc.abi.json
write:sc.manifest.json
SUCC
```

### CODE REVIEW

File: src/Neo.Compiler.MSIL/CmdOptions.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Neo.Compiler
+{
+    public class CmdOptions
+    {
+        [Option('f', "file", Required = true, HelpText = "Specify the file to compile to NEF file")]
+        public string Filename { get; set; }
+
+        [Option('r', "references", Required = false, HelpText = "Compilation references")]
+        public IEnumerable<string> References { get; set; }
```
Comment (lock9) [inline src/Neo.Compiler.MSIL/CmdOptions.cs]: Hi @shargon, can't we use `.csproj` file and get the references from there?

File: src/Neo.Compiler.MSIL/CmdOptions.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+        [Option('c', "compile", Required = false, HelpText = "Is the source already compiled?")]
+        public bool Compile { get; set; } = false;
```
Comment (lock9) [inline src/Neo.Compiler.MSIL/CmdOptions.cs]: I don't think this is necessary. Can't we use the file extension instead?

**Comment by lapetina9** [comment]
## Neo PR Test Report

**Repository:** neo-devpack-dotnet  
**Number:** 131  
**Title:** Allow neon to compile from the source  
**Change description:** Neon was updated to support *.cs file compilation.


## Code version / branch

**Neo-devpack-dotnet:**  shargon:compile-source

## Test Results

| Test Name | Network | OS | Test Result | Comments |
| ------- |:-------:| --:| -----------:| --------:|
| Initial UT |  | Windows 10 x64 | OK | 
| Compile Contract 1 from unit tests |  | Windows 10 x64 | OK | |
| Compile Contract 2 from unit tests |  | Windows 10 x64 | NOK | Conversion errors occurred during compilation. Failed identification of Smart Contract Framework method. |

## Additional info
We made two tests that should have the same results, but one of them failed to compile the smart-contract. We followed instructions provided in the PR.

unknown



## **Test List**

**1. Initial UT**  
Run solution unit tests.

**2. Compile Contract 1 from unit tests**  
  
The test was to compile the same file that is used in unit tests, using the command line. 
Contract code:
```CSharp
namespace Neo.Compiler.MSIL.TestClasses
{
    public class Contract1 : SmartContract.Framework.SmartContract
    {
        //default smartcontract entry point.
        //but the unittest can be init from anywhere
        //no need to add code in Main.
        public static object Main(string method, object[] args)
        {
            return UnitTest_001();
        }
        public static byte[] UnitTest_001()
        {
            var nb = new byte[] { 1, 2, 3, 4 };
            return nb;
        }

    }
}
```

Expected result:
- File compiled sucessfully.

**3. Compile Contract 2 from unit tests**  
The test was to compile the same file that is used in unit tests, using the command line. 
Contract code:
```CSharp
namespace Neo.Compiler.MSIL.TestClasses
{
    public class Contract2 : SmartContract.Framework.SmartContract
    {
        //default smartcontract entry point.
        //but the unittest can be init from anywhere
        //no need to add code in Main.
        public static object Main(string method, object[] args)
        {
            Neo.SmartContract.Framework.Services.Neo.Runtime.Notify(args[0]);
            Neo.SmartContract.Framework.Services.Neo.Runtime.Notify(args[2]);
            return UnitTest_002();
        }
        public static byte UnitTest_002()
        {
            var nb = new byte[] { 1, 2, 3, 4 };
            return nb[2];
        }
    }
}

```

Expected result:
- File compiled sucessfully.

Actual result:
Compilation failed caused by Neo.SmartContract.Framework not being found.

**Comment by shargon** [comment]
> Can we make the Neo.SmartContract.Framework to be automatically referenced?

If everyone want to, yes, it's easy

File: src/Neo.Compiler.MSIL/CmdOptions.cs (hunk 2 above)
Comment (shargon) [reply src/Neo.Compiler.MSIL/CmdOptions.cs]: Yes, it's a good idea

File: src/Neo.Compiler.MSIL/CmdOptions.cs (hunk 1 above)
Comment (shargon) [reply src/Neo.Compiler.MSIL/CmdOptions.cs]: I prefer to use the source... It's more versatile

**Comment by shargon** [comment]
> but the code is not working.

What doesn't work, could you explain it?

File: src/Neo.Compiler.MSIL/CmdOptions.cs (hunk 2 above)
Comment (shargon) [reply src/Neo.Compiler.MSIL/CmdOptions.cs]: Also I will add `.vb`

**Comment by shargon** [comment]
> Compilation failed caused by Neo.SmartContract.Framework not being found.

That's the reason of references 😂 You obtain the same behavior in Visual Studio hahahah

**Comment by lightszero** [comment]
I think build from a csproj is more better.

**Comment by shargon** [comment]
@lock9 could you review it again?

File: src/Neo.Compiler.MSIL/CmdOptions.cs (hunk 1 above)
Comment (shargon) [reply src/Neo.Compiler.MSIL/CmdOptions.cs]: Done, but it works with manual references too

**Comment by lock9** [comment]
@shargon we are trying to deploy a SC into neo 3 testnet but we are having several issues. I will test this later today

File: src/Neo.Compiler.MSIL/Program.cs [hunk 3]
```diff
@@ -50,27 +54,112 @@ public static void Main(string[] args)
...
+                    }
+                case ".vb":
+                    {
+                        // Compile VB source
+
+                        log.Log("Compiling from VB source");
+                        var output = Compiler.BuildVBScript(new string[] { args.Filename }, args.References.ToArray());
+                        fs = new MemoryStream(output.Dll);
+                        fspdb = new MemoryStream(output.Pdb);
+                        break;
```
Comment (lock9) [inline src/Neo.Compiler.MSIL/Program.cs]: Hi @shargon, my particular opinion is that we could encapsulate these instructions inside methods, like CompileCSProj, CompileCSFile or CompileDll.

File: src/Neo.Compiler.MSIL/CmdOptions.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+using CommandLine;
+using System.Collections.Generic;
+
+namespace Neo.Compiler
+{
+    public class CmdOptions
+    {
+        [Option('f', "file", Required = true, HelpText = "Specify the file to compile to NEF file")]
```
Comment (lock9) [inline src/Neo.Compiler.MSIL/CmdOptions.cs]: I don't think we need this parameter. Do we?

Comment (shargon) [reply src/Neo.Compiler.MSIL/CmdOptions.cs]: If we need references, we need to know which is the file

Comment (erikzhang) [reply src/Neo.Compiler.MSIL/CmdOptions.cs]: We don't need `-f`. We can detect the suffix name. If it's ".cs", then compile it from csharp source file. If it's ".csproj", then compile it from csharp project.

**Comment by erikzhang** [comment]
I think "Neo.SmartContract.Framework.dll" should be added as a reference by default. There is no need to add it manually.

File: src/Neo.Compiler.MSIL/CmdOptions.cs (hunk 4 above)
Comment (shargon) [reply src/Neo.Compiler.MSIL/CmdOptions.cs]: But if we want to specify references? if it is a dll?

File: src/Neo.Compiler.MSIL/Neo.Compiler.MSIL.csproj [hunk 5]
```diff
@@ -22,8 +22,13 @@
...
   <ItemGroup>
-    <PackageReference Include="Mono.Cecil" Version="0.11.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.3.1" />
+    <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic" Version="3.3.1" />
+    <PackageReference Include="Mono.Cecil" Version="0.11.1" />
     <PackageReference Include="Neo" Version="3.0.0-CI00212" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\Neo.SmartContract.Framework\Neo.SmartContract.Framework.csproj" />
```
Comment (lock9) [inline src/Neo.Compiler.MSIL/Neo.Compiler.MSIL.csproj]: @erikzhang should we use project reference or should we use nuget? @shargon  I suspect we need to use the nuget reference. Let's wait for Erik's response.

Comment (erikzhang) [reply src/Neo.Compiler.MSIL/Neo.Compiler.MSIL.csproj]: Using project reference is okay if they are in the same solution.

File: src/Neo.Compiler.MSIL/Program.cs [hunk 6]
```diff
@@ -50,27 +66,68 @@ public static void Main(string[] args)
...
+                    {
+                        // Compile VB files
+
+                        log.Log("Compiling from VB source");
+                        var output = Compiler.CompileVBFile(new string[] { fileInfo.FullName }, new string[0]);
+                        fs = new MemoryStream(output.Dll);
+                        fspdb = new MemoryStream(output.Pdb);
+                        break;
+                    }
+                default:
```
Comment (erikzhang) [inline src/Neo.Compiler.MSIL/Program.cs]: dll?

File: src/Neo.Compiler.MSIL/Program.cs [hunk 7]
```diff
@@ -50,27 +66,68 @@ public static void Main(string[] args)
...
-            ILModule mod = new ILModule(log);
-            Stream fs;
-            Stream fspdb = null;
-
-            //open file
-            try
+            switch (fileInfo.Extension.ToLowerInvariant())
             {
-                fs = File.OpenRead(filename);
+                case ".csproj":
```
Comment (erikzhang) [inline src/Neo.Compiler.MSIL/Program.cs]: Can't compile from "vbproj"?

File: src/Neo.Compiler.MSIL/Compiler.cs [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+        /// <summary>
+        /// Build script
+        /// </summary>
+        /// <param name="filename">File name</param>
+        /// <returns>Assembly</returns>
+        public static Assembly CompileVBProj(string filename)
+        {
+            ExtractFileAndReferences(filename, ".vb", out var files, out var references);
+            return CompileVBFile(files.ToArray(), references.ToArray());
```
[the remark this answers is not in the document]
Comment (shargon) [reply src/Neo.Compiler.MSIL/Compiler.cs]: this == vb?
. net compile the same IL code between languages, then.. the bug will be for both languages

**Pull request merged**